### ISSUE TITLE
React 16 Umbrella ☂️ (and 15.5)

### ISSUE THREAD

**Issue by gaearon** [facebook/react#8854]
This list might change, just putting it out there as the first draft.
Posting these together since they're related, and we don't plan more 15.x releases after 15.5 anyway.

To put these changes in context, we have a few goals:

(Click to show)
<details>

 * For the past several months, we have been working on a [rewrite of React codenamed “Fiber”](https://www.youtube.com/watch?v=aV1271hd9ew&feature=youtu.be). Initially, it won’t affect public API, but it brings several new features (like https://github.com/facebook/react/issues/2127 and https://github.com/facebook/react/issues/2461). Fiber gives us a solid foundation to improve React core in numerous ways. We’ll be talking more about it soon, and we intend to ship it with React 16 by default.

 * To reduce the bundle size, we need to remove the APIs that we don't recommend, such as `createClass` and `React.DOM.*` helpers. We intend to warn once about their usage so that you can start removing dependencies on them. They will still be available as separate packages, but we will exclude them from the default build of React 16.

* We would like to have more control over the bundles so that we can better optimize them. This is why we are considering switching to flat bundles (and thus removing access to React internals in `react/lib/*` and `react-dom/lib/*`) in React 16. This will also mean faster compile times by default for users of Webpack and other bundlers, and faster server-side rendering performance.

</details>

Here’s a speculative list of changes we think of doing in these releases:

# Past Releases

I’ve moved these to the end of this post.

# 16 Final

- [x] https://github.com/facebook/react/issues/10625 (@flarnie & @sophiebits)
- [x] Check if https://github.com/facebook/react/pull/10142 is a real regression (@nhunzaker, @aweary)
- [x] Check Google issue (<s>still exists after RC2</s> turns out to be polyfill.io bug, https://github.com/Financial-Times/polyfill-service/issues/1299) https://www.reddit.com/r/javascript/comments/6pqrz5/comment/dmrl0nh
- [x] Figure out what we want to do regarding https://github.com/facebook/react/pull/10612 and https://github.com/facebook/react/issues/10672 @gaearon
- [x] Ensure Enzyme isn't blocked on anything from our side for 16 compat. (@flarnie)
- [x] Remove `unstable_` prefix from `unstable_createPortal` and alias it @gaearon https://github.com/facebook/react/pull/10675
- [x] Ensure we show a sensible warning for this case https://github.com/facebook/react/pull/10673 @gaearon
- [x] Preparing for the Final Release: (React team in general)
  - [x] Write blog post adapted from #10294 (@acdlite)
    - [x] Add attribute info to big changes section
    - [x] Consolidate/remove "Updates" section at the end
    - [x] Server hydration: make hydration changes more prominent in breaking changes section, maybe write a full blog post later (https://github.com/facebook/react/issues/10591)
- [x] Release an update to DevTools with https://github.com/facebook/react-devtools/pull/892
- [x] Website (@bvaughn and @flarnie)
- [x] Docs 
  - [x] Document ability to return strings and arrays from `render`
  - [x] Document `createPortal()` 
  - [x] Merge [server streaming](https://github.com/facebook/react/blob/master/docs/docs/reference-react-dom-node-stream.md) doc into regular dom server doc and fix imports to match new unified entry point
  - [x] Test renderer https://github.com/facebook/react/pull/10692
- [x] Right after release:
  - [x] Update CDN URLs on website and official fiddle to reference new paths (`/umd/*` rather than `sebmt/*`)
  - [x] Update doc snippets (such as all codepens on the website!) to use 16.

*TODO: Add any bugs discovered while testing RC.*

# 16 Wishlist

- [x] Fix hot reloading in RN (probably needs `unstable_forceUpdateDeep()` and then change https://github.com/gaearon/react-deep-force-update to use it)
- [x] Show a better message for missing React (https://github.com/facebook/react/issues/10406)
- [x] Make sure we offer a way of canceling `setState` updater form (https://twitter.com/nishb1/status/852542873803345920) (@gaearon will follow up)
- [x] Remove Stack and compat code
- [x] Add support to Enzyme for Fiber (airbnb/enzyme#1007)

----

## Past Releases

<details>

# 15.5

(Click to show)

<details>

- [x] ReactDOM fixes, if @jquense @nhunzaker @aweary want to release any.
- [x] We should probably ship a fix for https://github.com/facebook/react/issues/8487 (https://github.com/facebook/react/pull/9005)
- [x] New warning: callback refs on stateless functional components (https://github.com/facebook/react/pull/8635, https://github.com/facebook/react/pull/8739)
- [x] New API deprecations:
  - [x] `React.createClass` (will be moved to a separate package, [and we provide a codemod](https://github.com/reactjs/react-codemod#explanation-of-the-new-es2015-class-transform-with-property-initializers) if you want to update to ES classes)
  - [ ] ~~`React.createFactory` (will be moved to a separate package or removed, it's a [one-liner](https://github.com/facebook/react/blob/eb89bc4c301c5970b1ec88d3fdfa2202d4dca4f8/src/isomorphic/classic/element/ReactElement.js#L271-L272))~~
  - [ ] <s>`React.DOM.*` factories (will be moved to a separate package, https://github.com/facebook/react/pull/8356)</s> **We forgot to do it. Moving to 15.6**
  - [ ] <s>`React.createMixin` (will be removed, https://github.com/facebook/react/pull/8853)</s> **We forgot to do it. Moving to 15.6**
  - [x] `React.propTypes` (will be moved to a separate package)
  - [x] Deprecate React addons https://github.com/facebook/react/issues/9207
  - [x] Decide how to handle all of this with the UMD addons build
  - [x] Fix this bug https://github.com/facebook/react/issues/9351

</details>

# 15.6

See https://github.com/facebook/react/issues/9398.

# 16 Alpha

(Click to show)

<details>

- [x] Remove deprecations added in 15
  - [x] Stop appending `px` to strings (https://github.com/facebook/react/pull/6899)
  - [x] Remove `LinkedStateMixin`, `valueLink` and `checkedLink` (https://github.com/facebook/react/pull/8165)
  - [x] Remove `React.__spread` and `React.createMixin`
  - [x] Move `React.createClass` to a separate package
  - [x] Move `PropTypes` to a separate package.
  - [x] Move `React.DOM.*` factories to a separate package
  - [x] Decide what to do with the addons build
  - [x] If we keep any addons, they need to work regardless of whether React is aliased to a flat bundle or not (which is not the case now)
- [x] Strip PropTypes checkers in production build (https://github.com/facebook/react/pull/8066)
- [x] Re-add warning about calling PropTypes directly (https://github.com/facebook/react/issues/8080)
- [X] More ReactDOM fixes, if @jquense @nhunzaker @aweary want to release any.
- [x] Make a decision on the input fix by @jquense (for now accepted, but follow up work will change this. @flarnie link to issue on related follow up work)

</details>

# 16 Beta

(Click to show)

<details>

- [x] Actually enable fragments (@flarnie verified this)
- [x] Make error boundaries official (@bvaughn)
  - [x] Choose lifecycle naming
  - [x] Codemod stuff
  - [x] Decide whether to runtime warn for the old name (decided: let's not do it)
- [x] Fix the issue with `invokeGuardedCallback` (tracking in https://github.com/facebook/react/issues/9577, @acdlite is working on this)
- [x] (This came up internally) Set a hard limit for recursion https://github.com/facebook/react/issues/9193. (@acdlite)
- [x] Duplicate React + string refs results in very confusing error (ref is not a function). We should provide a good invariant for that. (https://github.com/facebook/react/issues/9962, @spicyj @flarnie and @gaearon) (is this Draft issue also about this? https://github.com/facebook/draft-js/issues/296)
  - [x] Is there a test for this? (@flarnie)
- [x] Warn (throw?) when doing an update on a container that was manually emptied outside of React (in stack, this mounted a brand-new tree; in fiber, it tries to apply an update and usually fails) (@flarnie in https://github.com/facebook/react/pull/10210)
  - [x] Is there a test for this? (@flarnie)
- [x] Additional APIs we need in order to start experimenting with async. (@acdlite)
  - [x] `React.unstable_AsyncComponent`
  - [x] `ReactDOM.flushSync(batch)`
- [x] sanity test - use in a CRA app, and check fixtures
- [x] Messaging
  - [x] Update isfiberreadyyet.com
  - [x] Compose a Tweet

</details>

# 16 RC

(Click to show)
<details>

- [x] [m-l] Pass DOM props through (https://github.com/facebook/react/issues/10399, latest is: https://github.com/facebook/react/pull/10416) (@gaearon, @spicyj, @sebmarkbage, @flarnie, @acdlite)
- [x] Regression: we should warn if object is passed to an event listener prop #10407 (@aweary) (fixed? in https://github.com/facebook/react/pull/10453 but needs <s>review</s> updated and merged)
- [x] Make sure input variables (props, state) point to the correct values for each lifecycle. Example where this isn't the case for `componentWillMount`: https://jsfiddle.net/m3pL3yaj/ (@acdlite) (fixed in #10481)
- [x] `select`'s onChange event fires with incorrect selectedIndex, value (https://github.com/facebook/react/issues/10420 @sebmarkbage)
- [x] How do we ensure weak minifiers don't ship the bundle twice? https://github.com/facebook/react/issues/9589 (@flarnie, @trueadm) (fixed in https://github.com/facebook/react/pull/10446)
- [x] Bug: SSR has false positive warnings for SVG tags #10415 (@spicyj) (fixed in https://github.com/facebook/react/pull/10452)
- [x] Bug: Unexpected cross-domain error passed to `componentDidCatch` #10441 (@bvaughn) (fixed in https://github.com/facebook/react/pull/10447)
- [x] Warn when nesting a 15 tree inside 16 (@spicyj https://github.com/facebook/react/pull/10434) (fixed in https://github.com/facebook/react/pull/10434)
- [x] undefined is not a function (evaluating 'owner.getName()') https://github.com/facebook/react/issues/10443
- [x] Regressions reported in beta
  - [x] The server renderer crashes in production with inline styles. (https://github.com/facebook/react/issues/10299, fixed by https://github.com/facebook/react/pull/10300)
  - [x] The server renderer doesn't yet support returning arrays and strings from components (fixed by #10221)
  - [x] The server renderer still renders `data-reactid` somewhat unnecessarily. (https://github.com/facebook/react/issues/10306) (@gaearon)
    - [x] Potentially fixed by https://github.com/facebook/react/pull/10339 (@gaearon)
  - [x] In some cases `Error: null` is reported instead of the real error. (https://github.com/facebook/react/issues/10321) [decision: update the error message] (@acdlite)
  - [x] Shallow renderer doesn't implement `unstable_batchedUpdates`. [decision: team seems to be leaning away from doing this; discussion in #10355]
- [x] [s-m] `shouldComponentUpdate` with functional components [decision: delete it for now, add it in a minor] (@sebmarkbage) https://github.com/facebook/react/pull/10371
- [x] There is a report that Google crawler can’t render the page using 16 ([link](https://www.reddit.com/r/javascript/comments/6pqrz5/react_16_beta/dkrgvzj/?utm_content=permalink&utm_medium=front&utm_source=reddit&utm_name=javascript)). (@flarnie)
- [x] <s>Check if https://github.com/facebook/react/issues/10405 is a real regression</s> (No, it's been fixed looks like.)
- [x] [s-m] Ensure type validation didn't regress when we enabled new features https://github.com/facebook/react/issues/9577 (@gaearon)
- [x] <s>Decide if we want to break value attribute syncing behavior (https://github.com/facebook/react/pull/10150)</s> [decided to postpone until 17]
- [x] [s-m] Decide if we want to stop calling `componentDidUpdate` from shallow renderer (https://github.com/facebook/react/pull/10372)
- [x] Decide: what do we do about polyfills?
  - [x] Map/Set (need to throw early, @gaearon) https://github.com/facebook/react/pull/10356
  - [x] Object.assign (keep or assume it's there?) https://github.com/facebook/react/pull/10280 [decision: keep it bundled for now]
  - [x] rAF [decision: keep it required]
(Maybe use the same approach/standard as Relay? https://facebook.github.io/relay/docs/relay-modern.html#javascript-environment-requirements)
- [x] Big missing pieces
  - [x] Add server rendering and reviving
  - [x] Add shallow renderer
    - [x] We seem to be decoupling shallow renderer from ReactDOM. However in 15 `findDOMNode()` works inside of shallow renderer because it contains ReactDOM injections. Is this going to be a breaking change? (@gaearon) [decision: let's not support this]
- [x] Decide on whether to include warning whitelisting/blacklisting (@bvaughn) [decision: not in 16.0]
- [x] Maybe: Switch to flat bundles (no more `react-dom/lib/*`, internals are truly private)
  - [x] Decide: plan forward for popular projects depending on internals (`react-native-web`, `react-tap-event-plugin`) (might not block final)
     - [x] `EventPluginHub` is "dangerously" exported from flat bundle for tap event plugin
     - [x] `react-native-web` probably needs more internals, coordinate with @necolas (done in https://github.com/facebook/react/pull/10138)
- [x] <s>Support `class` and `for` https://github.com/facebook/react/pull/10169 (@gaearon)</s> [decision: not gonna do this now]
- [x] Feature parity for renderers
  - [x] Shallow renderer supports strings and arrays
  - [x] Server renderer supports strings and arrays
  - [x] <s>Server renderer supports error boundaries</s> [doesn't make sense conceptually right now]
- [x] Decide on whether to put `react-current-owner` in its own package to reduce issues caused by npm duplication [not now]
- [x] Decide on DOMServer entry points (`react-dom/server` and `react-dom/node-stream` seem inconsistent with each other. Why is only one of them `server`?) (@gaearon) https://github.com/facebook/react/pull/10362
  * These were configured separately for Prepack. One target includes both the pure JS APIs and the Node APIs and the other only includes the pure JS APIs. We could combine them into a single `react-dom/server` entry point and add a new `renderToString` method on it.
- [x] Decide: Do we care about catching errors in events in the initial release? [no]
- [x] Error messages (@flarnie) (Dan: I moved this back to blockers because I don't think we can release with unclear messages. Error boundaries are a huge change in open source and we must have documentation and a link before we start unmounting roots.)
  - [x] "You can add an error boundary" message should link to docs
  - [x] duplicate key warning for children with same keys is out of date; claims child will be ignored but now we will only render second one. (easy to add to 16 final or sooner) (@sebmarkbage)
- [x] Verify that the bundles produced are valid in strict mode. https://github.com/facebook/react/issues/10294#issuecomment-319490960 (@bvaughn)
- [x] Testing the Beta: (React team in general)
  - [x] Share the beta and proposed release date with library authors several weeks in advance
  - [x] Triage any issues reported with beta.
</details>
</details>

**Comment by aweary** [comment]
> ReactDOM fixes, if @jquense @nhunzaker @aweary want to release any.

I think at least https://github.com/facebook/react/pull/8575 and https://github.com/facebook/react/pull/7359 should be included in `15.5`. They're both close to being merged.

**Comment by nhunzaker** [comment]
@jquense we probably want to merge yours (first  #8575) first. #7359 relies on ChangeEventPlugin to keep the value attribute in sync with the value property. We'll need to re-evaluate #7359 with those changes.

**Comment by thejameskyle** [comment]
> - React.createClass will be moved to a separate package
> - React.createFactory will be moved to a separate package
> - React.DOM.* factories will be moved to a separate package

All these new packages, maybe you need a [tool](https://lernajs.io/) for managing so many packages in a single repo... :wink:

**Comment by spicyj** [comment]
As long as it doesn't make our UMD builds harder I'm on board with lerna. Someone want to send a PR?

**Comment by gaearon** [comment]
Pretty sure [we've reached peak harder](https://github.com/facebook/react/blob/046f7448d7224da5b33441248a32a319a6464bc5/grunt/config/browserify.js#L76-L97) and it's downhill from there.

**Comment by thejameskyle** [comment]
@gaearon Challenge accepted.

@spicyj jk I won't make it harder, I'll throw something together for you

**Comment by tizmagik** [comment]
This is awesome! Excited about what's ahead in React 16. Also glad I won't have to manually create a minified build of React for server render. 😅

Curious, any idea about how much larger the bundle is with React Fiber?

**Comment by brenoc** [comment]
Why is this scratched?

> Make setState async by default.
> Remove the attribute whitelist.
> Stateful functional components.
> New context API.

**Comment by jquense** [comment]
Ya! look great y'all. one thought related to flat bundles. it's likely to write certain types of tooling without access to internal apis. specifically thinking about test libs, the few I maintain invariably need to touch a lot of internal apis when test utils fall short. it's a bit of a niche case tho as the cost of using those Apis is well understood and aren't used in production apps

**Comment by yaycmyk** [comment]
If you guys do want to switch to lerna, I've been working on rollup scripting to make flat files of lerna packages. Would be happy to help out.

**Comment by gaearon** [comment]
>Curious, any idea about how much larger the bundle is with React Fiber?

Right now it's roughly the same size while implementing more features (async scheduling, error boundaries, fragments, portals). We haven't worked on optimizing its size so I think we can reduce it further too. We'll likely do so after the initial release.

Another promising (IMO) area is Fiber's support for custom renderers. If ReactDOM's tradeoffs don't match what your product needs, it should be much easier to create a custom DOM renderer that, for example, doesn't ship with the synthetic event system. Fiber makes custom renderers *way* easier to create, while giving you all component features out of the box.

**Comment by Hurtak** [comment]
'Stateful functional components' is there some issue where I can read more about this?

**Comment by swernerx** [comment]
When you go the flat bundle route - which I definitely appreciate - it would be awesome to offer a modern bundle for Node v6 / ES2015 capable browsers, too.

I did some work on this topic in http://npmjs.com/package/prepublish - which effectively produces flat bundles for different output scenarios. Basically a configured wrapper around Rollup/Babel/Buble.

I also thought of adding specific "modern" entries to the package.json so that bundles like Webpack can make use of this e.g. `main:modern` or `module:modern`. I think something like this would be definitely interesting for React (especially in context of SSR apps where you can target Node >= v6).

**Comment by Kitanotori** [comment]
Agree with @swernerx . It's about time to migrate to ES6.

**Comment by gaearon** [comment]
@swernerx This is an interesting direction and we may explore it in the future. Right now the browser support for ES6 is increasing but it many features are sufficiently slower than ES5 counterparts, just because nobody had time to optimize it yet. I know V8 team is hard at work fixing this, and I’d rather look into it again after evergreen browsers are mostly fast when using ES6.

**Comment by gaearon** [comment]
>'Stateful functional components' is there some issue where I can read more about this?

No specific issue AFAIK. There are some old ideas in [this repo](https://github.com/reactjs/react-future) if you’re interested, but they’re just that, ideas. There are numerous reasons we’re interested in having stateful functional components, [I listed some here](https://twitter.com/dan_abramov/status/823818032942379008).

**Comment by dtinth** [comment]
Concerning the “flat bundles,” it would be great if the internals are still exposed one way or the other, because we at Taskworld sometimes monkey-patched React internals from userland code to have more fine-grained error messages and performance measurement.

What we did is we hooked into React’s internals and added some instrumenting code like this:

<details><summary>Show code</summary>

```js
  const ReactReconciler = require('react/lib/ReactReconciler')

  ReactReconciler.mountComponent = ((mountComponent) => function (internalInstance) {
    enter('React mount ' + describeInstance(internalInstance))
    try { return mountComponent.apply(this, arguments) } finally { exit() }
  })(ReactReconciler.mountComponent)

  ReactReconciler.receiveComponent = ((receiveComponent) => function (internalInstance) {
    enter('React update ' + describeInstance(internalInstance))
    try { return receiveComponent.apply(this, arguments) } finally { exit() }
  })(ReactReconciler.receiveComponent)

  ReactReconciler.unmountComponent = ((unmountComponent) => function (internalInstance) {
    enter('React unmount ' + describeInstance(internalInstance))
    try { return unmountComponent.apply(this, arguments) } finally { exit() }
  })(ReactReconciler.unmountComponent)
```
</details>

…

We also hooked into Redux and Reselect in similar ways for a unified development experience.

I know that this isn’t supported and internal APIs may break in any release (especially with Fiber coming), but unsupported is better than impossible 😄.

**Comment by gaearon** [comment]
>we at Taskworld sometimes monkey-patched React internals from userland code to have more fine-grained error messages

This is supported by error boundaries in Fiber.

>and performance measurement.

You can use ReactPerf for that.

>I know that this isn’t supported and internal APIs may break in any release (especially with Fiber coming), but unsupported is better than impossible 😄.

Even better, let’s discuss the use cases and come up with supported APIs. 😉

**Comment by swernerx** [comment]
@gaearon It's about offering such a package format for modern clients (in parallel to the classic build - which is what the prepublish tool mentioned does). In the end React does not decide on the processing/bundling/output of the application. But if you transpile stuff you make the original ES6 code unaccessible for any user who wants to deploy for modern node or embedded browsers only.

**Comment by swernerx** [comment]
@gaearon "main" and "module" should always point to the ES5 bundles with either commonjs or esmodule outputs. But we are free to offer an alternative modern build keeping ES2015 stuff intact.

**Comment by MilosRasic** [comment]
Excuse my ignorance, but does removal of createMixin mean that all packages that rely on mixins will stop working until dependants switch to a HOC solution or will they be able to quickly switch to an alternate way to provide old-style mixins?

**Comment by gaearon** [comment]
>Excuse my ignorance, but does removal of createMixin mean that all packages that rely on mixins will stop working until dependants switch to a HOC solution or will they be able to quickly switch to an alternate way to provide old-style mixins?

`createMixin` implementation looks like this:

```js
function createMixin(mixin) {
  return mixin;
}
```

It was never documented, and only appeared in the API because we thought we might want to validate mixins in the future. Packages that somehow found out about it and started using it can just remove the `createMixin` call and use the mixin directly.

Most packages providing mixins don’t use `createMixin`. For those that do, this is a one-line change: just removing the useless function call.

>It's about offering such a package format for modern clients (in parallel to the classic build - which is what the prepublish tool mentioned does)

Yes, I understand this. What I’m saying is if we offer this build, people *will* use it, and if performance is suboptimal, they *will* blame React. So we will wait until ES6 features are actually fast before providing a build that does that.

**Comment by Kitanotori** [comment]
@gaearon Migration to ES6 can be done in iterations, e.g., first convert all `var`s to either `let` or `const`, then other features that are supported without performance issues.

Also, there are two issues here: (1) syntax of the source code, and (2) syntax of the bundle. Is there a reason for not migrating the source to the latest standard - ES7/ES2016 - with the possibility of using staged features if deemed useful enough?

Syntax of the bundle is a totally separate issue, as the syntax of the bundle (or bundles) can be decided during the build step. I think it would be awesome to have separate bundles for each relevant standard, but I'm not sure how to achieve this in practice, or whether this can be achieved without unreasonable amount of work. Nevertheless, ES6 is here to stay, and the project ought to provide bundle targeting it, which naturally first requires converting the source to utilize its features.

**Comment by gaearon** [comment]
We are pragmatic in choosing both source and target languages, and are driven by pragmatic needs.

>Migration to ES6 can be done in iterations, e.g., first convert all vars to either let or const, then other features that are supported without performance issues.

For the target bundle, what exactly would that accomplish? There are no benefits to the end users in this case, and there are downsides (less browser support, more fragmentation and potential surface area for bugs).

>Is there a reason for not migrating the source to the latest standard - ES7/ES2016 - with the possibility of using staged features if deemed useful enough?

We *are* using new features in newly written code. There is very little benefit, in our opinion, to rewrite existing code with new features since this also introduces a risk of new bugs. It’s a nice way to ramp up contributors, but as noted in this issue, *we have already rewritten most of React* so it doesn’t make sense to me to convert the old code now.

**Comment by Kitanotori** [comment]
@gaearon I'm not sure if you read my whole comment. It doesn't matter to the user whether the source is ES5, ES6, or ES7 - only the bundle matters. If only ES5 bundle is to be provided, so be it, but why continue using obsolete syntax for the source when we have transpilation tools available?

I noticed that there actually seems to be old and new syntax already side-by-side in the code: part of the codebase using only `var`s and `require`s, some part using `const`s, `let`s, and `import`s (to name some of the basic differentiating features). May I suggest cleaning the codebase of obsolete syntax for the sake of clarity and maintainability?

**Comment by zalmoxisus** [comment]
@gaearon 

> Even better, let’s discuss the use cases and come up with supported APIs.

It would be great to have an API to detect if an object is a Synthetic Event. For Redux, usually users are passing events as action payloads ([which also `redux-actions` does](https://github.com/zalmoxisus/redux-devtools-extension/issues/287)), even when not using them. It breaks the ability to access them while sending to the monitors. A simple solution was:

```js
import SyntheticEvent from 'react/lib/SyntheticEvent';
// ...

const store = createStore(rootReducer, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__({
  serialize: {
    replacer: (key, value) => {
      if (value && value instanceof SyntheticEvent) return '[Event]';
      return value;
    }
  }
}));
```

Though maybe we should just [encourage using `event.persist` when not providing event data explecitely](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/Troubleshooting.md#it-fails-to-serialize-data-when-passing-synthetic-events-or-calling-an-action-directly-with-redux-actions).

**Comment by gaearon** [comment]
>Though maybe we should just encourage using event.persist when not providing event data explecitely.

This is the solution here.

**Comment by Kitanotori** [comment]
@gaearon I totally understand if code that may soon be deprecated is not rewritten, but is it really a good idea to, for example, continue using two module syntaxes side-by-side in _newly written code_? I see some source files even combining both. For parts that are necessary for supporting Node, using Node's module syntax is of course understandable. Personally I feel quite strongly about unnecessarily mixing two competing module handling mechanisms - particularly in newly written code, and particularly when talking about a standard- and a non-standard one. 

One point, not sure if actually relevant, came to my mind: React is one of the "hottest" libraries out there in the JavaScript world, and if React decided to fully utilize the latest standardized ES features (preferably with strict style guidelines), this might incentivize browser manufacturers to hasten their optimization efforts. React could show way for, not only other UI libraries out there, but for the whole JavaScript ecosystem. Maybe slightly optimistic, but 😺

**Comment by gaearon** [comment]
>using two module syntaxes side-by-side in newly written code? I see some source files even combining both. 

We are using CommonJS exclusively right now (mostly due to Facebook build pipeline limitations). I think you might be confusing Flow's `import type` syntax with ES modules.

**Comment by Kitanotori** [comment]
Oh, haven't used [Flow](https://flowtype.org/) so that kind of syntax was new for me.

**Comment by dtinth** [comment]
@gaearon Thanks for the suggestions. I am excited about the error boundaries!

I find off-the-shelf solutions (such as ReactPerf) sometimes not to be suitable. For instance, I monkeypatched several libraries to be able to construct a table that shows how much time (in percentage) is spent in:

- React reconciling doing DOM manipulations (`react-dom/lib/ReactReconciler`)
- Userland React lifecycle (`react-dom/lib/ReactInstrumentation`)
- Selectors created using `reselect`
- The Redux store reducer
- The Redux store subscriber

Having this data gives a high level overview of where I should focus on optimizing.

However, let’s not focus on a specific problem (performance, error boundaries)—these are just examples. I believe there are other unforeseen use cases where monkeypatching the internals could be useful.

Here’s another situation where I monkeypatched React’s internals: I want to make React warnings appear as a popup in the DOM instead of in the console, because sometimes I develop my user interface without the devtools on.

I accomplished this by monkeypatching `fbjs/lib/warning` to display a popup instead of using `console.error()`. This would be very hard to accomplish if React is packaged as a flat bundle—I would have given up that idea. Exposing the internals allows for experimentation.

**Comment by gaearon** [comment]
>I find off-the-shelf solutions (such as ReactPerf) sometimes not to be suitable. For instance, I monkeypatched several libraries to be able to construct a table that shows how much time (in percentage) is spent in: [...]

I may be wrong but I believe all this data is already exposed via ReactPerf.getLastMeasurements. 

>I want to make React warnings appear as a popup in the DOM instead of in the console, because sometimes I develop my user interface without the devtools on.

So do we which is why we intend to show a UI (https://github.com/facebook/react/pull/7360) by default in React 16. I forgot to add this to the umbrella. 

> let’s not focus on a specific problem (performance, error boundaries)—these are just examples. I believe there are other unforeseen use cases where monkeypatching the internals could be useful.

I agree it *can* be useful. However the problem is, once you've monkeypatched it to your convenience, you won't come back filing an issue and gathering support for it from the community. As a result everyone ends up with local hacks that break every few versions, and no permanent solutions, while the maintainers aren't even aware this is something people want. I would rather prefer we have more discussions like this one, where you explain the use cases, and we either point out or provide (possibly unstable) hooks.

**Comment by dtinth** [comment]
@gaearon That makes sense, and I agree with your sentiment. Although I usually treat open-source projects with the “you want it, you do it” mindset (which is why I kinda like to start by experimenting alone), making them an issue could bring more benefit to the project and community. ❤️

**Comment by MHerszak** [comment]
I can see my browser console showing so many warnings just because of `react-router`. :)

**Comment by gaearon** [comment]
We likely won’t be firing multiple warnings, just a single one (with names of the first few classes). We don’t want to spam the console.

**Comment by guotie** [comment]
when does it release

**Comment by gaearon** [comment]
We don't know yet but I'd say likely within three months.

**Comment by lambert-velir** [comment]
+1 for 'Stateful functional components'.  I'm not sure what the API would look like, but it has to be more elegant than es6 classes.

**Comment by lax4mike** [comment]
+1 for Stateful functional components.  I don't know what the API would look like, but it has to be more elegant than es6 classes.

**Comment by voidale** [comment]
that's really awesome cant wait to get Fiber means we can finally improve the renderToString to achieve better SSR performance? Async renderToString will be the best thing for me right now went from 2 servers to 10 just because it's not async.

**Comment by gaearon** [comment]
Fiber doesn't currently bring anything new for SSR.

The plan mentions that *if we switch to flat bundles*, the synchronous `renderToString` perf should get better *by default* because you won't have to envify your code manually (and most people don't). However this is not related to Fiber or SSR being asynchronous.

**Comment by ColCh** [comment]
What about stateless components optimization aka function components? They are pure, and they got rendered every time, not like [Classes with PureRenderers](https://facebook.github.io/react/docs/pure-render-mixin.html)

Definition of stateless component is here [c/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L38](https://github.com/facebook/react/blob/3bc5595dfd6db741b19f17e7aef98b2c1e4be847/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L38). It looks like regular component. May be it should look like [React.PureComponent](https://facebook.github.io/react/docs/react-api.html#react.purecomponent) ?

Sorry if I don't understand some points of stateless components

**Comment by MilosRasic** [comment]
To be honest, stateful functional components seems to be just another way to write what you can already write using classes. I'd be far more hyped about stateless functional component optimizations that have been hinted at in the past.

**Comment by gaearon** [comment]
@ColCh You can refer to the past discussion about this which was pretty controversial: .https://github.com/facebook/react/pull/6914

**Comment by unirey** [comment]
Fiber, Relay 2.. React Conf '17 is probably going to be a sad day for me because I'll hype myself to cosmos, unless..

**Comment by unirey** [comment]
Fiber, Relay 2.. React Conf '17 is probably going to be a sad day for me because I'll hype myself to cosmos.

**Comment by phaistonian** [comment]
ETA for 16?

**Comment by gaearon** [comment]
In a few months.

**Comment by aweary** [comment]
@gaearon what is the status of PropTypes? We have [previously stated](https://github.com/facebook/react/pull/6780#issuecomment-219463952) that they were in "maintenance mode" and turned away feature requests, but we haven't been consistent about that. Do we plan to continue adding features to PropTypes?

**Comment by gaearon** [comment]
@aweary I don't think so, but do you have any particular ones in mind?

**Comment by aweary** [comment]
@gaearon no, I'm mostly asking in regards to the handful of PRs we have open for PropTypes (on mobile, can't link).

**Comment by aweary** [comment]
@gaearon a few I found are: https://github.com/facebook/react/pull/8823, https://github.com/facebook/react/pull/8684, https://github.com/facebook/react/pull/6938

**Comment by danielearwicker** [comment]
How about moving state/setState into a separate library? Core React becomes props-only. Statefulness can then be added on in various ways, with "per component state" as just one option. Derived class such as `StatefulComponent` would serve as base class for any component wanting intrinsic state.

**Comment by asci** [comment]
Could somebody explain me please, why `createClass` is not recommended? In favor of future Stateful functional components?

**Comment by gaearon** [comment]
@danielearwicker 

Let's keep this thread focused on 16 roadmap and not use it as a place for generic feature requests or suggestions. Otherwise everyone will pile on the features they want to see which have no relationship to the roadmap. 😉

**Comment by danielearwicker** [comment]
@gaearon sure, though in my defence I'm asking for fewer features 😎

**Comment by gaearon** [comment]
State is first-class citizen in React, see https://github.com/facebook/react/issues/3398#issuecomment-98850930 for some thoughts about it.

**Comment by aweary** [comment]
@gaearon should we add removing `isMounted` and `replaceState` from class components to the 16 roadmap? They are currently deprecated.

**Comment by yaycmyk** [comment]
Why is removing `isMounted()` a thing? It's useful for async callbacks. It's a very simple function, barely adds any size to the library, and implementing similar functionality as the user means opting in to 2 separate callbacks (cDM and cWU).

**Comment by aweary** [comment]
@yaycmyk see https://github.com/facebook/react/issues/5465 which links to a number of issues providing background on why it was deprecated.

**Comment by cvle** [comment]
I see the ReactCSSTransitionGroup addon will soon be deprecated. I've been working on a similar project for some time now and it needs some exposure. Check it out here: [react-css-transition](https://wikiwi.github.io/react-css-transition/).

**Comment by gaearon** [comment]
@cvle Do you want to join forces with @taion and @jquense on https://github.com/reactjs/react-transition-group?

**Comment by cvle** [comment]
@gaearon I'm open for collaboration. Though both projects are quiet different at the moment and I have no idea in which way the project https://github.com/reactjs/react-transition-group is heading, while https://github.com/wikiwi/react-css-transition is almost feature complete (though it needs some testing on the battlefield).

You know what, I'm gonna open an issue over at https://github.com/reactjs/react-transition-group to discuss this.

**Comment by koba04** [comment]
I have a question.
React deprecates `React.createClass` in v15.5 and removes it in v16.
Is a separate package for `React.createClass` (`react-create-class`?) released at the time of v15.5?
Is the separate package in `facebook/react` repo or in `reactjs` organization like `react-transition-group`?

**Comment by salmankhann** [comment]
@gaearon can this issue also be part of 15.5? 
https://github.com/facebook/react/issues/7211

**Comment by SanthoshHari912** [comment]
Any plan to revive the licensing of React library ? Right now the Patents added along with the BSD licensing has become a debatable topic, and many enterprises back foot on using React for the same reason.

**Comment by gaearon** [comment]
@salmankhann We’re not sure. See https://github.com/facebook/react/pull/8575#issuecomment-284192820 for latest thinking on this.

@SanthoshHari912 I don’t know if you’ve seen other threads about this, but I believe everything we have to say about this is already there, including who to reach out to if you still want to argue about this. I would appreciate if technical threads were not derailed with this because it is very demotivating.

**Comment by koba04** [comment]
I just noticed that `Move PropTypes to a separate package.` was added to this list.
What about PropTypes used in Context as `contextTypes` and `childContextTypes`?
Are there any plans for changing Context API?

**Comment by probablyup** [comment]
I'm curious if propType checking is really necessary for context. It seems like merely opting in to a key would be sufficient and simplify the API.

**Comment by acdlite** [comment]
@koba04 @probablyup We have plans to rethink context but it won't happen in the near future.

**Comment by koba04** [comment]
@acdlite Thanks! I got it.

In v16, Do we use PropTypes and Context like this?
(We need to install the separated package for PropTypes when using Context)

```js
import React from 'react';
import PropTypes from 'react-prop-types'; // or something

const Box = ({text}, {theme}) => <div className={theme}>{text}</div>;
Box.propTypes = {
  text: PropTypes.string
};
Box.contextTypes = {
  theme: PropTypes.string
};
```

**Comment by gaearon** [comment]
Posted a plan for addons: https://github.com/facebook/react/issues/9207

**Comment by jochenberger** [comment]
15.5.0-rc.1 prints a warning about `ReactCSSTransitionGroupChild` using `React.createClass`. Maybe reconsider #7701 for 15.5+?

**Comment by jochenberger** [comment]
And also one about `React.PropTypes` being used. Unfortunately, there is no location information printed along with the warning, but I was able to trace it down to `ReactTransitionGroup`.
I guess these warnings appear because I use the react-with-addons UMD package. If the addons package is also deprecated, maybe there should be a warning when the file is loaded.

**Comment by jochenberger** [comment]
Or maybe the best solution would be to somehow suppress those warnings from within the addons build. If you need to use the addons build (e.g. because you want to use `React.Perf`), there's not really an alternative in 15.5.

**Comment by gaearon** [comment]
Good point, it's indeed confusing that we warn in our own build. I agree we should suppress these warnings for our own uses of them.

The addons build *will be* deprecated because each individual addon is deprecated. TestUtils is moving to ReactDOM, and ReactPerf will not be available in 16 (instead you can use component profiling in browser DevTools Timeline). But I'm not sure if it makes sense to have an additional warning about the build itself or not.

**Comment by jochenberger** [comment]
It would probably make more sense to show individual warnings for the deprecated addons if people use them.

**Comment by gaearon** [comment]
Looks like there's a serious bug in 15.5 RC: https://github.com/facebook/react/issues/9351

**Comment by jochenberger** [comment]
Shouldn't `prop-types` 15.5.0 come with an UMD build?

**Comment by gaearon** [comment]
It should. Also I think it makes sense to move doc to its README.
cc @acdlite @flarnie

**Comment by jochenberger** [comment]
The {{prop-types}} topic is confusing (at least to me). The docs link to https://github.com/aackerman/PropTypes, but that doesn't have any tags and the last commit was over a year ago. https://www.npmjs.com/package/prop-types however links to https://github.com/facebook/react and shows, that "15.5.0 is the latest of 4 releases ".

**Comment by gaearon** [comment]
Yea I don't understand it either yet 😛 we'll clear it up with the release.

**Comment by bvaughn** [comment]
FYI the serious bug in RC1 (#9351) was resolved with an RC2 release yesterday evening.

> The docs link to https://github.com/aackerman/PropTypes, but that doesn't have any tags and the last commit was over a year ago.

@jochenberger: Sorry if I'm overlooking something, but which docs link to this project?

> https://www.npmjs.com/package/prop-types however links to https://github.com/facebook/react and shows, that "15.5.0 is the latest of 4 releases ".

This is the correct package to use going forward. If we have any links in our docs that point to the other one, I'll be happy to correct them.

**Comment by gaearon** [comment]
>If we have any links in our docs that point to the other one, I'll be happy to correct them.

I think this line was related to this from #9359:

https://github.com/facebook/react/pull/9359/files#diff-cfcea6c43d62e54e6c86684cd87e2f5dR8

**Comment by gaearon** [comment]
>The React team recently took over the prop-types package.

What's the plan for where the source of truth will be?

**Comment by jquense** [comment]
did ya'll still want `react-prop-types` or do I not need to worry about transferring that anymore?

**Comment by gaearon** [comment]
We ended up on `prop-types` I guess. Since it's not very React specific and could potentially be used by Preact, Inferno, and something else.

**Comment by acdlite** [comment]
@jquense We're using `prop-types` as the source of truth.

One scenario I'm worried about is folks installing `react-prop-types` without realizing it's not the official package. For example, in addition to `create-react-class` we also own `react-create-class` so we can point users to the correct package.

**Comment by gaearon** [comment]
@acdlite Where will source code for `prop-types` live? Who will manage the repository?

**Comment by acdlite** [comment]
@gaearon Right now it's just in our 15.5-dev branch. Haven't decided where it should live. I'm cool with either leaving it in the React repo for now, or moving it to its own repo. Any preference?

**Comment by acdlite** [comment]
And I think we should continue to maintain it, since PropTypes remain popular and there's no great alternative yet. (There's Flow and TypeScript, but I think we agree that's not a viable option for every team.)

**Comment by gaearon** [comment]
`reactjs/prop-types`?

**Comment by gaearon** [comment]
The reason I'd like to move it out is to stop flooding React issue tracker with PropTypes feature requests. They get lost here and don't get the necessary discussion.

**Comment by acdlite** [comment]
Sounds good to me

**Comment by bvaughn** [comment]
👍  for reactjs/prop-types

**Comment by coodoo** [comment]
Is there any update on the optimizations for `stateless` functional component since now v16/fiber is right at the corner?

**Comment by necolas** [comment]
On the topic of `react-dom/lib/*` modules…I wrote up some notes about the problems with the proposed approach and using `ResponderEventPlugin` on web.

> the Responder modules…have non-trivial dependencies on other React internals that aren't exposed in the public release of `react-dom`. I don't think forking the Responder modules is a good solution

More: https://github.com/necolas/react-native-web/issues/364#issuecomment-294243735

**Comment by probablyup** [comment]
@gaearon doesn't prop-types error out if you try to use it outside of a React context?

**Comment by gaearon** [comment]
@probablyup No if you use the supported API for checking it (`PropTypes.checkPropTypes`).

**Comment by gaearon** [comment]
>Is there any update on the optimizations for stateless functional component since now v16/fiber is right at the corner?

They take a separate path in Fiber so should have slightly less overhead. Nothing dramatic though.

**Comment by LucaColonnello** [comment]
@gaearon can I suggest something about this one `Make a decision on shouldComponentUpdate with functional components`?

Recently I did an experiment on writing virtual dom without libraries in 300 lines, starting from an already done implementation.

https://jsfiddle.net/Luca_Colonnello/5xebu97u/3/

So in this lite version, I used `shouldComponentUpdate` attached to nodes, wherever they are and it turned out I was able to override it from the parent of a component. I found it pretty useful and easy.

```js
const shouldCommentsListUpdate = (props, nextProps) => {
   return props.comments !== nextProps.comments;
};

const List = ({ items }) => (
<div className='list'>
   {items.map((item) => (<div className='listItem'>{item}</div>))}
</div>
);

const Comments = ({ comments }) => (
<div className='commentSection'>
   <List items={comments} shouldUpdate={shouldCommentsListUpdate} />
</div>
);
```

This gives us the ability to overwrite the inner implementation based on parent rules as well as apply the `shouldUpdate` function to other node which are not components.

Let me know what do you think about this solution! :)

**Comment by jochenberger** [comment]
Is there a repository for `create-react-class`? The UMD module does not work (React is undefined in line 764) and I'd like to report an issue, but I'm not sure where I'm supposed to do that.

**Comment by gaearon** [comment]
Please file an issue in this repo. The module itself is in the 15-stable branch (since we don't plan further development).

**Comment by unirey** [comment]
Im trying to get my hands dirty but I can't find any information if Fiber is the default renderer for `@next` or if there is a flag or something.

Im using `16.0.0-alpha.12` and I get an error when I try to return several components. Is Stack still default renderer or this feature is not implemented yet?

**Comment by bvaughn** [comment]
> Im trying to get my hands dirty but I can't find any information if Fiber is the default renderer for `@next` or if there is a flag or something.

Fiber is the only renderer actually exported in `react-dom@next`.

**Comment by vicbell** [comment]
React v16 is not release until now, why?

**Comment by rashidul0405** [comment]
@vicbell because it's not ready yet :)

**Comment by Diar77** [comment]
Anyone knows when React 16 gets released?

**Comment by bvaughn** [comment]
> Anyone knows when React 16 gets released or have an idea?

The team hasn't yet announced (or settled on) a release date.

> Also will it be easy to migrate an existing create-react-app to React 16 when it releases?

Yes, that's the plan. One of the primary purposes of the 15.5 and 15.6 releases was to add deprecation warnings to prepare people for 16.0.

> And is it worth using React 16 instead of 15 if you are creating a streaming movie website like netflix?

It should be fine to use 15 for now and upgrade once 16 is released. Pay attention to any dev-mode warning messages in 15 about things that have been deprecated and you should have no problems.

**Comment by flarnie** [comment]
Just updated this with remaining open issues from Phases 6.1-6.3 of https://github.com/facebook/react/issues/7925 and I'm archiving that issue. We can still reference the open items for Phase 7-9 when planning future work for Fiber and future releases.

**Comment by rchanou** [comment]
FYI, these sections are misformatted. Fixed below.

# 15.5

<details>

- [x] ReactDOM fixes, if @jquense @nhunzaker @aweary want to release any. 
- [x] We should probably ship a fix for https://github.com/facebook/react/issues/8487 (https://github.com/facebook/react/pull/9005) 
- [x] New warning: callback refs on stateless functional components (https://github.com/facebook/react/pull/8635, https://github.com/facebook/react/pull/8739) 
- [x] New API deprecations: 
- [x] `React.createClass` (will be moved to a separate package, [and we provide a codemod](https://github.com/reactjs/react-codemod#explanation-of-the-new-es2015-class-transform-with-property-initializers) if you want to update to ES classes) 
- [ ] ~~`React.createFactory` (will be moved to a separate package or removed, it's a [one-liner](https://github.com/facebook/react/blob/eb89bc4c301c5970b1ec88d3fdfa2202d4dca4f8/src/isomorphic/classic/element/ReactElement.js#L271-L272))~~ 
- [ ] `React.DOM.*` factories (will be moved to a separate package, https://github.com/facebook/react/pull/8356) **We forgot to do it. Moving to 15.6** 
- [ ] `React.createMixin` (will be removed, https://github.com/facebook/react/pull/8853) **We forgot to do it. Moving to 15.6** 
- [x] `React.propTypes` (will be moved to a separate package) - [x] Deprecate React addons https://github.com/facebook/react/issues/9207 - [x] Decide how to handle all of this with the UMD addons build 
- [x] Fix this bug https://github.com/facebook/react/issues/9351

</details>

# 16 Alpha

<details>

- [x] Remove deprecations added in 15 
- [x] Stop appending `px` to strings (https://github.com/facebook/react/pull/6899) 
- [x] Remove `LinkedStateMixin`, `valueLink` and `checkedLink` (https://github.com/facebook/react/pull/8165) 
- [x] Remove `React.__spread` and `React.createMixin` 
- [x] Move `React.createClass` to a separate package 
- [x] Move `PropTypes` to a separate package. 
- [x] Move `React.DOM.*` factories to a separate package 
- [x] Decide what to do with the addons build 
- [x] If we keep any addons, they need to work regardless of whether React is aliased to a flat bundle or not (which is not the case now) 
- [x] Strip PropTypes checkers in production build (https://github.com/facebook/react/pull/8066) 
- [x] Re-add warning about calling PropTypes directly (https://github.com/facebook/react/issues/8080) 
- [X] More ReactDOM fixes, if @jquense @nhunzaker @aweary want to release any. 
- [x] Make a decision on the input fix by @jquense (for now accepted, but follow up work will change this. @flarnie link to issue on related follow up work)

</details>

**Comment by flarnie** [comment]
Thanks @rchanou - correct me if I'm missing something, but it looks like you remove the indentations on the checkbox lists and the words '(Click to show)' above the 'Details' dropdowns.

The indentations were intentional - the indented items are related to the previous un-indented item, but maybe we didn't make the relationship clear enough.

I think the words '(Click to show)' were added because of confusion when people saw the 'Details' dropdown. I don't know how common it is to use the `<details>` tag on Github and people may not be used to it.

I'm happy to make changes to fix any other formatting issues, and thanks again!

**Comment by rchanou** [comment]
@flarnie, since the original post is fixed I hid my "fix" (which I realize needed some fixing itself 😅). Thanks!

**Comment by iddan** [comment]
When will I be able to return strings in component's render? As we use the 16 alpha w/ RN I thought this is already possible

**Comment by gaearon** [comment]
RN does not use Fiber yet. The "react" package is not very relevant here because it doesn't contain the reconciler logic. It only has helpers like createElement and Component. So it's not Fiber. We'll switch RN to Fiber some time soon but don't have an exact date.

**Comment by gaearon** [comment]
16 is out.
Remaining items moved to https://github.com/facebook/react/issues/11088.
